### DISCUSSION TITLE
How to use IKFast for new robot type?

### DISCUSSION THREAD

**Discussion by fishbotics** [tesseract-robotics/tesseract#777] [General]
Hi all,

I'm just getting started using Tesseract and I'd like to be able to use it for the Franka Panda. In the past, I've used IKFast (because, well, it's  fast), but I had to generate a new C++ file using OpenRave. In the ikfast directory, I don't see any robot-specific info. Can someone point me to how I might be able to integrate a new robot type into the IK solver?

Thanks!

**Comment by Levi-Armstrong** [comment]
Take a look [here](https://github.com/tesseract-robotics/tesseract/tree/master/tesseract_kinematics/test) at the following files.
- abb_irb2400_ikfast_kinematics.h
- abb_irb2400_ikfast_kinematics.cpp
- abb_irb2400_ikfast_solver.hpp.

Also there is iiwa7 example in the same directory.

**Comment by Levi-Armstrong** [comment]
After you create the kinematics object you will need to create the [factory](https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_kinematics/kdl/src/kdl_factories.cpp) which is the plugin to tesseract.

**Comment by fishbotics** [comment]
Thanks @Levi-Armstrong. I might be misunderstanding here, but that factory you linked is for KDL. I see in [here](https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_kinematics/test/kinematic_plugins.yaml) that you list all the plugins. The iiwa seems to point to KDL--not IKFast and abb points to OPW, which I think is defined [here](https://github.com/tesseract-robotics/tesseract/blob/8445d507a2bbe3ec5b5e819a835a0d33758cc13b/tesseract_kinematics/opw/include/tesseract_kinematics/opw/opw_factory.h).

Is the OPW IK library your preferred standard? I'm used to using IKFast, but I'd be happy to set up something else if it's better. Or, do you have an example of an IKFast factory?

**Comment by gavanderhoorn** [comment]
If possible, try to use OPW.

The 'only' thing you'd need is the parameters. The plugin code is already written, and it's both faster and much more compact than IKFast.

Whether the Panda can be modelled as an OPW-type robot I don't know.

**Comment by fishbotics** [comment]
![image](https://user-images.githubusercontent.com/45679966/166986307-47eff1f9-768d-4d7e-bc68-ad1a236b8fd6.png)

It looks pretty similar to the OPW model, but I don't know how to get all the parameters. Maybe I can load it in sim and figure out how to take the measurements.

**Comment by Levi-Armstrong** [comment]
> Thanks @Levi-Armstrong. I might be misunderstanding here, but that factory you linked is for KDL. I see in [here](https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_kinematics/test/kinematic_plugins.yaml) that you list all the plugins. The iiwa seems to point to KDL--not IKFast and abb points to OPW, which I think is defined [here](https://github.com/tesseract-robotics/tesseract/blob/8445d507a2bbe3ec5b5e819a835a0d33758cc13b/tesseract_kinematics/opw/include/tesseract_kinematics/opw/opw_factory.h).
> 
> Is the OPW IK library your preferred standard? I'm used to using IKFast, but I'd be happy to set up something else if it's better. Or, do you have an example of an IKFast factory?

@fishbotics Here are the ikfast examples for the iiwa

https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_kinematics/test/iiwa7_ikfast_kinematics.h
https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_kinematics/test/iiwa7_ikfast_kinematics.cpp

Generated from ikfast
https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_kinematics/test/iiwa7_ikfast_solver.hpp

**Comment by marip8** [comment]
OPW kinematics can only be used with a very specific robot configuration, and I don't think this particular robot meets the criteria. A few general requirements are that the robot must be 6 DOF and have a spherical wrist (i.e. the last three axes all intersect at a single point). [Here is a picture of a valid OPW configuration with the dimensions denoted](https://github.com/Jmeyer1292/opw_kinematics/blob/3c66f52b26279cf04c157efe15d8c37ada8e49c1/opw.png). If your robot doesn't have an exact correlation to this diagram, then OPW can't be used

**Comment by marip8** [comment]
I'm coincidentally also trying to implement a custom IKFast plugin currently, and it seems like the functionality for loading a custom plugin is fairly limited in one case and broken in another. 

## Case 1: YAML config file (limited capability)
You can specify the names and locations of plugins in the plugin config file, such as [this example](https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_kinematics/test/kinematic_plugins.yaml#L2-L5). However, I believe these names have to be hard-coded, so they wouldn't transfer between machines unless the plugins were installed to a universal directory (which is not the case for most ROS users building in a workspace in their home directory).

## Case 2: Environment variables (broken)
You can also specify the names and locations of plugins using [two (fairly hidden) environment variables](https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_kinematics/core/src/kinematics_plugin_factory.cpp#L35-L36). A user could use a script or ROS launch file to populate these environment variables in a generic way that is machine independent.

However, the plugin loader does not attempt to load plugins that it finds from environment variables. Instead it only loads libraries from its internal cache (see below that `plugins_local` never really gets used). I think the fix for this is simple: iterate over `plugins_local` rather than `this->search_libraries`.

https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_common/include/tesseract_common/plugin_loader.hpp#L133-L162

Are there other ways of loading custom plugins, or do we need to fix these methods? 

Also the kinematic plugin unit tests don't test this loading capability because the plugin loader already knows about the plugins built by `tesseract` through the [`TESSERACT_KINEMATICS_PLUGINS` compiler definition](https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_kinematics/core/src/kinematics_plugin_factory.cpp#L50-L51).

**Comment by marip8** [comment]
I guess there is a third case which is to use some combination of both other methods, specifically setting the search path in the environment variable and adding the plugin library name to the YAML config file. We still should address the shortcomings of the other approaches too though

**Comment by Levi-Armstrong** [comment]
> ## Case 1: YAML config file (limited capability)
> You can specify the names and locations of plugins in the plugin config file, such as [this example](https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_kinematics/test/kinematic_plugins.yaml#L2-L5). However, I believe these names have to be hard-coded, so they wouldn't transfer between machines unless the plugins were installed to a universal directory (which is not the case for most ROS users building in a workspace in their home directory).

I don't think that is the case. It first searches the directories but then reverts to system location and what is in your environment variables. I usually don't add anything here and as long as the terminal is sourced it will find them.

Update: 
To clarify, you do need to add the library to the list but really do not need to add a search path.

**Comment by Levi-Armstrong** [comment]
> ## Case 2: Environment variables (broken)
> You can also specify the names and locations of plugins using [two (fairly hidden) environment variables](https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_kinematics/core/src/kinematics_plugin_factory.cpp#L35-L36). A user could use a script or ROS launch file to populate these environment variables in a generic way that is machine independent.
> 
> However, the plugin loader does not attempt to load plugins that it finds from environment variables. Instead it only loads libraries from its internal cache (see below that `plugins_local` never really gets used). I think the fix for this is simple: iterate over `plugins_local` rather than `this->search_libraries`.

The environment variable was added to support containerization like using snap. Other than that it should really never need to be used because it searches the system location for the library.

**Comment by Levi-Armstrong** [comment]
> However, the plugin loader does not attempt to load plugins that it finds from environment variables. Instead it only loads libraries from its internal cache (see below that plugins_local never really gets used). I think the fix for this is simple: iterate over plugins_local rather than this->search_libraries.

This piece may very well be broke and will need to look at it more closely.

**Comment by Levi-Armstrong** [comment]
@marip8 Did you run into a case where it was unable to find the plugin?

**Comment by marip8** [comment]
> > ## Case 1: YAML config file (limited capability)
> > You can specify the names and locations of plugins in the plugin config file, such as [this example](https://github.com/tesseract-robotics/tesseract/blob/eebe3fd4570bf4c4a0cfc548ce0299d6b2a3c3df/tesseract_kinematics/test/kinematic_plugins.yaml#L2-L5). However, I believe these names have to be hard-coded, so they wouldn't transfer between machines unless the plugins were installed to a universal directory (which is not the case for most ROS users building in a workspace in their home directory).
> 
> I don't think that is the case. It first searches the directories but then reverts to system location and what is in your environment variables. I usually don't add anything here and as long as the terminal is sourced it will find them.
> 
> Update: To clarify, you do need to add the library to the list but really do not need to add a search path.

For some reason this doesn't seem to work on a ROS2 system after I have sourced my workspace and added the name of the library to the config file. It shows a search path in my tools workspace where `tesseract` was built (which I think actually comes from the target compile definition from `tesseract_kinematics`?), but the current workspace where the custom plugin is built does not have a known search path

```
snp_motion_planning_node-6] Warning: Failed to load symbol 'MotomanHC10InvKinFactory'
[snp_motion_planning_node-6]          at line 310 in /home/mripperger/ros/snpd/snpd_tools_ws/src/tesseract/tesseract_kinematics/core/src/kinematics_plugin_factory.cpp
[snp_motion_planning_node-6] Error:   Failed to instantiate plugin 'MotomanHC10InvKinFactory', Details: 
[snp_motion_planning_node-6] Search Paths (Search System Folders: True):
[snp_motion_planning_node-6]     - /home/mripperger/ros/snpd/snpd_tools_ws/install/tesseract_kinematics/lib
[snp_motion_planning_node-6] Search Libraries:
[snp_motion_planning_node-6]     - libsnp_motion_planning_plugins.so
[snp_motion_planning_node-6]     - libtesseract_kinematics_core_factories.so
[snp_motion_planning_node-6]     - libtesseract_kinematics_kdl_factories.so
[snp_motion_planning_node-6]     - libtesseract_kinematics_opw_factory.so
[snp_motion_planning_node-6]     - libtesseract_kinematics_ur_factory.so

```

**Comment by Levi-Armstrong** [comment]
Is `libsnp_motion_planning_plugins.so` location in your LD_LIBRARY_PATH environment variable?

**Comment by Levi-Armstrong** [comment]
I also think it would be good to update the parsing of the search path to use the resource locator if possible so you may use a package url.

**Comment by Levi-Armstrong** [comment]
Here is the documentation for what [dlopen](https://man7.org/linux/man-pages/man3/dlopen.3.html) searches.

**Comment by Levi-Armstrong** [comment]
@marip8 Is this library by chance in a pure cmake package?

**Comment by marip8** [comment]
> Is this library by chance in a pure cmake package?

No, it's in a ROS2 package.

> Is libsnp_motion_planning_plugins.so location in your LD_LIBRARY_PATH environment variable?

Actually no. Turns out it was a few ROS2 errors. I forgot to add an `ament_export_targets` command, so the package location wasn't added to the `LD_LIBRARY_PATH`. I also didn't realize that ROS2 libraries need to be installed in `lib` not `lib/<package_name>`. Once I fixed those, it works as expected

**Comment by Levi-Armstrong** [comment]
@fishbotics Were you able to integrate your ikfast?

**Reply by fishbotics** [reply]
@Levi-Armstrong , sorry it's taken me so long to get back to you on this. I was crunching for an unrelated deadline and had to put everything else aside for a bit.

I'm just getting back to this now and will be working on this for the next few days. I'll update here as I work on getting it figured out.

**Comment by marip8** [comment]
@Levi-Armstrong would you be interested in adding [this IKFast factory](https://github.com/ros-industrial-consortium/scan_n_plan_workshop/blob/3a4ebb5014034774f74e43687a40d9d1247fc85d/snp_motion_planning/src/motoman_hc10_ikfast_plugin.cpp) as a boilerplate file in `tesseract_planning`? It should be generically applicable to all IKFast plugins (minus the HC10 include and plugin export). Users could then make their own `tesseract` IKFast kinematics plugins in 3 lines in a `.cpp` file.

```c++
#include <tesseract_kinematics/ikfast/ikfast_factory_boilerplate.hpp>  // proposed addition to tesseract_kinematics
#include "custom_ikfast_generated_code.hpp"
TESSERACT_ADD_PLUGIN(tesseract_kinematics::IKFastInvKinFactory, MyCustomIKFastInvKinFactory)
```

**Reply by Levi-Armstrong** [reply]
Sounds great to me.

**Reply by rjoomen** [reply]
Implemented and submitted as #916.
